src/main/java/rx/internal/operators/OperatorConcat.java

```diff
@@ -212,6 +227,11 @@ public void onError(Throwable e) {
         @Override
         public void onCompleted() {
             if (once.compareAndSet(0, 1)) {
+                // signal the production count at once instead of one by one
+                long c = produced;
+                if (c != 0L) {
```

Comment:
Move the `c!=0` test into the `produced` method? Don't see why you need to assign `produced` to `c` as it can't be changed by another thread due Observable contract.